.github/workflows/docker-build.yml

```diff
@@ -0,0 +1,172 @@
+name: Docker Build
+
+on:
+  pull_request:
+    branches: [ main ]
+    paths:
+      - '**/Dockerfile'
+
+permissions:
+  contents: read
+  packages: write
+  pull-requests: read
+  id-token: write
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  detect-changes:
+    runs-on: ubuntu-latest
+    outputs:
+      base_matrix: ${{ steps.set-matrix.outputs.base_matrix }}
+      clients_matrix: ${{ steps.set-matrix.outputs.clients_matrix }}
+      protocols_matrix: ${{ steps.set-matrix.outputs.protocols_matrix }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      
+      - name: Get changed Dockerfiles
+        id: changed-files
+        uses: tj-actions/changed-files@v40
+        with:
+          files: |
+            **/Dockerfile
+
+      - name: Generate build matrices
+        id: set-matrix
+        run: |
+          # Get list of changed Dockerfiles
+          CHANGED_FILES=$(echo "${{ steps.changed-files.outputs.all_changed_files }}" | tr ' ' '\n')
+          
+          # Create matrices using jq
+          echo "base_matrix=$(echo "$CHANGED_FILES" | grep "^node-base/" | sed 's|/Dockerfile$||' | jq -Rsc 'split("\n")[:-1] | {include: map({image_path: .})}')" >> $GITHUB_OUTPUT
+          echo "clients_matrix=$(echo "$CHANGED_FILES" | grep "^clients/" | sed 's|/Dockerfile$||' | jq -Rsc 'split("\n")[:-1] | {include: map({image_path: .})}')" >> $GITHUB_OUTPUT
+          echo "protocols_matrix=$(echo "$CHANGED_FILES" | grep "^ethereum/" | sed 's|/Dockerfile$||' | jq -Rsc 'split("\n")[:-1] | {include: map({image_path: .})}')" >> $GITHUB_OUTPUT
+
+  build-base:
+    needs: detect-changes
+    if: ${{ fromJson(needs.detect-changes.outputs.base_matrix).include[0] }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix: ${{ fromJson(needs.detect-changes.outputs.base_matrix) }}
+    steps:
+      - uses: actions/checkout@v4
+      - name: Log in to GitHub Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Generate version
+        id: version
+        run: |
+          SHA=$(git rev-parse --short HEAD)
+          echo "image_name=node-base" >> $GITHUB_OUTPUT
+          echo "image_tag=${SHA}" >> $GITHUB_OUTPUT
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Build and push base image
+        uses: docker/build-push-action@v5
+        with:
+          context: ./${{ matrix.image_path }}
+          push: true
+          build-args: |
+            GRAFANA_LOKI_API_KEY=${{ secrets.GRAFANA_LOKI_API_KEY }}
+            GRAFANA_PROM_API_KEY=${{ secrets.GRAFANA_PROM_API_KEY }}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          tags: ghcr.io/blockjoy/${{ steps.version.outputs.image_name }}:${{ steps.version.outputs.image_tag }}
```

Comment:
I think we should also include a latest tag as we previously discussed.  Trying to update all the dependencies with the hash is going to create a lot of work for every image update.  Or we omit the hash and stick to the semantic version (i.e. we shouldn't ever have a version 1.1.1-hask8d and a 1.1.1-ja9dk1...two of the same sematic with different hashes)

Reply:
Agreed, this mainly stemmed from the fact that I needed a way to bootstrap the pipeline since initially, no images were built that could satisfy the requirements of the next images. I ended up going for the commit of the hash and then updated the dependent Dockerfiles manually. We should only need the semver and only go for full release versioning, so don't allow pre-release versions on prod.
I'm not yet sure where we should draw the line between automations and manual deployments, especially to prod. Namely, we'll have renovate update Dockerfiles when new images get build (this is still something I need to test, haven't done it yet) and then CI/CD will pick up on those changes and built the three groups of images (node-base <-- clients <-- protocols) and (either renovate or CI/CD, TBD) will update the docker-container:// URI in the babel.yaml, but do we want afterwards to deploy the image to the API? I tend to think we do, so that it will become available on the frontend and then we could select what nodes should run what images independently